txpool/txpool.go

```diff
@@ -823,466 +531,508 @@ func (t *TxPool) validateTx(
+	defer func() {
+		a.promoted.unlock()
+		a.enqueued.unlock()
+	}()
+
+	currentNonce := a.getNonce()
+	if a.enqueued.length() == 0 ||
+		a.enqueued.peek().Nonce > currentNonce {
+		// nothing to promote
+		return 0
+	}
+
+	promoted := uint64(0)
+	nextNonce := a.enqueued.peek().Nonce
+	for {
+		tx := a.enqueued.peek()
+		if tx == nil ||
+			tx.Nonce != nextNonce {
+			break
+		}
 
-// max price ordered tx heap implementation
-type maxTxPriceHeapImpl struct {
-	txPriceHeapImplBase
-}
+		// pop from enqueued
+		tx = a.enqueued.pop()
+
+		// push to promoted
+		a.promoted.push(tx)
 
-func newMaxTxPriceHeapImpl() heap.Interface {
-	return &maxTxPriceHeapImpl{
-		txPriceHeapImplBase: txPriceHeapImplBase{
-			make([]*pricedTx, 0),
-		},
+		// update counters
+		nextNonce += 1
+		promoted += 1
 	}
-}
 
-func (t maxTxPriceHeapImpl) Less(i, j int) bool {
-	if t.txs[i].from == t.txs[j].from {
-		return t.txs[i].tx.Nonce < t.txs[j].tx.Nonce
+	// only update the nonce map if the new nonce
+	// is higher than the one previously stored.
+	if nextNonce > currentNonce {
+		a.setNonce(nextNonce)
 	}
 
-	return t.txs[i].price.Cmp(t.txs[j].price) >= 0
+	return promoted
 }
 
-type minTxPriceHeapImpl struct {
-	txPriceHeapImplBase
+// Lookup map used to find transactions present in the pool
+type lookupMap struct {
+	sync.RWMutex
+	all map[types.Hash]*types.Transaction
 }
 
-// min price ordered tx heap implementation
-func newMinTxPriceHeapImpl() heap.Interface {
-	return &minTxPriceHeapImpl{
-		txPriceHeapImplBase: txPriceHeapImplBase{
-			make([]*pricedTx, 0),
-		},
+// Adds the given transaction to the map. [thread-safe]
+func (m *lookupMap) add(txs ...*types.Transaction) {
+	m.Lock()
+	defer m.Unlock()
+
+	for _, tx := range txs {
+		m.all[tx.Hash] = tx
 	}
 }
 
-func (t minTxPriceHeapImpl) Less(i, j int) bool {
-	if t.txs[i].from == t.txs[j].from {
-		return t.txs[i].tx.Nonce < t.txs[j].tx.Nonce
-	}
+// Removes the given transactions from the map. [thread-safe]
+func (m *lookupMap) remove(txs ...*types.Transaction) {
+	m.Lock()
+	defer m.Unlock()
 
-	return t.txs[i].price.Cmp(t.txs[j].price) < 0
+	for _, tx := range txs {
+		delete(m.all, tx.Hash)
+	}
 }
 
-type localAccounts struct {
-	accounts map[types.Address]bool
-	mutex    sync.RWMutex
-}
+// Returns the transaction associated with the given hash. [thread-safe]
+func (m *lookupMap) load(hash types.Hash) (*types.Transaction, bool) {
+	m.RLock()
+	defer m.RUnlock()
 
-func newLocalAccounts(addrs []types.Address) *localAccounts {
-	accounts := make(map[types.Address]bool, len(addrs))
-	for _, addr := range addrs {
-		accounts[addr] = true
-	}
-	return &localAccounts{
-		accounts: accounts,
-		mutex:    sync.RWMutex{},
+	tx, ok := m.all[hash]
+	if !ok {
+		return nil, false
 	}
+
+	return tx, true
 }
 
-func (a *localAccounts) containsAddr(addr types.Address) bool {
-	a.mutex.RLock()
-	defer a.mutex.RUnlock()
-	return a.accounts[addr]
+// Gauge for measuring pool capacity in slots
+type slotGauge struct {
```

Comment:
You can move the slotGauge struct and all associated methods (& the slotsRequired fn) to a separate slot_gauge.go file

Reply:
[Exported](https://github.com/0xPolygon/polygon-sdk/pull/287/commits/8f8c2a793b02c4744e33f3598b871ea3b1aa73e9)